File: javascript/tests/after-refactor/IntegerArgumentMarshaler.test.js
const sinon = require('sinon');
const { ErrorCode } = require('../../after-refactor/ArgsError');
const { IntegerArgumentMarshaler } = require('../../after-refactor/IntegerArgumentMarshaler');

describe('#IntegerArgumentMarshaler', () => {
  describe('#set', () => {
    describe('Given a valid iterator', () => {
      it('Should read value from the iterator', () => {
        const mockIter = {
          next: sinon.stub(),
        };
        mockIter.next.returns({ done: false, value: '10' });
        const sut = new IntegerArgumentMarshaler();

        sut.set(mockIter);
        const value = IntegerArgumentMarshaler.getValue(sut);

        expect(mockIter.next.calledOnce).toBe(true);
        expect(value).toBe(10);
      });
    });

    describe('Given a iterator with bad int format', () => {
      it('Should throw an error', () => {
        const mockIter = {
          next: sinon.stub(),
        };
        mockIter.next.returns({ done: false, value: 'bad' });
        const sut = new IntegerArgumentMarshaler();

        let err;
        try {
          sut.set(mockIter);
        } catch (e) {
          err = e;
        }

        expect(mockIter.next.calledOnce).toBe(true);
        expect(err.getErrorCode()).toBe(ErrorCode.INVALID_INTEGER);
      });
    });

    describe('Given a iterator with no more value', () => {
      it('Should throw an error', () => {
        const mockIter = {
          next: sinon.stub(),
        };
        mockIter.next.returns({ done: true });
        const sut = new IntegerArgumentMarshaler();

        let err;
        try {
          sut.set(mockIter);
        } catch (e) {
          err = e;
        }

        expect(mockIter.next.calledOnce).toBe(true);
        expect(err.getErrorCode()).toBe(ErrorCode.MISSING_INTEGER);
      });
    });
  });

  describe('#getValue', () => {
    describe('Given a marshaler without calling set', () => {
      it('Should return the default value: 0', () => {
        const marshaler = new IntegerArgumentMarshaler();

        const value = IntegerArgumentMarshaler.getValue(marshaler);

        expect(value).toBe(0);
      });
    });
    describe('Given a non-marshaler object', () => {
      it('Should return the default value: 0', () => {
        const value = IntegerArgumentMarshaler.getValue(undefined);

        expect(value).toBe(0);
      });
    });
  });
});

File: java/src/test/java/args/after/IntegerArgumentMarshalerTest.java
package args.after;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.fail;

import java.util.Iterator;
import java.util.NoSuchElementException;
import org.testng.annotations.Test;

/**
 * Unit test for IntegerArgumentMarshaler.
 */
public class IntegerArgumentMarshalerTest {
  @Test
  @SuppressWarnings(value = "unchecked")
  public void set_validInput() throws Exception {
    Iterator<String> mockIterator = (Iterator<String>) mock(Iterator.class);
    when(mockIterator.next()).thenReturn("10");
    when(mockIterator.hasNext()).thenReturn(true);
    IntegerArgumentMarshaler sut = new IntegerArgumentMarshaler();

    sut.set(mockIterator);
    int value = IntegerArgumentMarshaler.getValue(sut);

    assertEquals(value, 10, "Should return 10 after the marshaler reads from the iterator");
    verify(mockIterator, times(1).description("Should only read one element from the iterator"))
        .next();
  }

  @Test
  @SuppressWarnings(value = "unchecked")
  public void set_invalidInput() throws Exception {
    Iterator<String> mockIterator = (Iterator<String>) mock(Iterator.class);
    when(mockIterator.next()).thenReturn("bad");
    when(mockIterator.hasNext()).thenReturn(false);
    IntegerArgumentMarshaler sut = new IntegerArgumentMarshaler();

    testThrowArgsException(sut, mockIterator, ArgsErrorCode.INVALID_INTEGER);
    verify(mockIterator, times(1).description("Should only read one element from the iterator"))
        .next();
  }

  private void testThrowArgsException(IntegerArgumentMarshaler marshaler, Iterator<String> input,
      ArgsErrorCode expectedCode) {
    try {
      marshaler.set(input);
    } catch (ArgsException e) {
      assertEquals(e.getErrorCode(), expectedCode);
      return;
    }
    fail(String.format("Expecting exception with code [%s], but got none", expectedCode.name()));
  }

  @Test
  @SuppressWarnings(value = "unchecked")
  public void set_noMoreInput() throws Exception {
    Iterator<String> mockIterator = (Iterator<String>) mock(Iterator.class);
    when(mockIterator.next()).thenThrow(new NoSuchElementException());
    when(mockIterator.hasNext()).thenReturn(true);
    IntegerArgumentMarshaler sut = new IntegerArgumentMarshaler();

    testThrowArgsException(sut, mockIterator, ArgsErrorCode.MISSING_INTEGER);
    verify(mockIterator, times(1).description("Should only read one element from the iterator"))
        .next();
  }

  @Test
  public void getValue_beforeCallingSet() throws Exception {
    IntegerArgumentMarshaler sut = new IntegerArgumentMarshaler();

    int value = IntegerArgumentMarshaler.getValue(sut);

    assertEquals(value, 0, "Should return 0 before the marshaler takes any input");
  }

  @Test
  public void getValue_nullInput() throws Exception {
    int value = IntegerArgumentMarshaler.getValue(null);

    assertEquals(value, 0, "Should return 0 given no marshaler");
  }

  @Test
  public void getValue_invalidMarshaler() throws Exception {
    ArgumentMarshaler unknownMarshaler = new ArgumentMarshaler() {
      @Override
      public void set(Iterator<String> currentArgument) throws ArgsException {
        currentArgument.hasNext();
      }
    };
    int value = IntegerArgumentMarshaler.getValue(unknownMarshaler);

    assertEquals(value, 0, "Should return 0 given unknown marshaler");
  }
}
